backend/hitas/views/apartment.py

```diff
@@ -141,8 +142,74 @@ def validate(self, data):
         if start > end:
             raise ValidationError({"start": "'shares.start' must not be greater than 'shares.end'."})
 
+        apartment_filter = Q(building__real_estate__housing_company__uuid=F("_housing_company"))
+        if self.parent.instance is not None:
+            apartment_filter &= ~Q(uuid=self.parent.instance.uuid)
+
+        building_id: Optional[str] = None
+        building_input = self.parent.initial_data["building"]
+        if isinstance(building_input, dict):
+            building_id = building_input.get("id")
+
+        if not isinstance(building_id, str) or not valid_uuid(building_id):
+            return data
+
+        apartments: Iterable[dict[str, Any]] = (
+            Apartment.objects.annotate(
+                _housing_company=Subquery(
+                    HousingCompany.objects.filter(real_estates__buildings__uuid=building_id).values("uuid"),
+                ),
+            )
+            .filter(apartment_filter)
+            .only("street_address", "apartment_number", "share_number_start", "share_number_end")
+            .values("street_address", "apartment_number", "share_number_start", "share_number_end")
+        )
```

Comment:
While this is working as expected, I don't like that checking for overlapping share numbers are done in Python (requires us to save ALL of the Housing Company's apartments to memory, and loop over them), instead we could query for overlapping share numbers directly from the database.

e.g.:
```python
## Validate overlapping share numbers
# If updating, exclude the apartment being created
apartment_filter = Q()
if self.parent.instance is not None:
    apartment_filter = ~Q(uuid=self.parent.instance.uuid)

building_id = self.parent.initial_data["building"].get("id")
housing_company_uuid = (
    HousingCompany.objects.filter(real_estates__buildings__uuid=building_id).only("uuid").first().uuid
)

overlapping_apartments = (
    Apartment.objects.filter(apartment_filter)
    .filter(building__real_estate__housing_company__uuid=housing_company_uuid)
    .filter(
        Q(share_number_start__lte=start, share_number_end__gte=start)  # Overlapping start
        | Q(share_number_start__lte=end, share_number_end__gte=end)  # Overlapping end
        | Q(share_number_start__gte=start, share_number_end__lte=end)  # Fully inside range
    )
)
```

Then, when returning the overlapping apartment addresses, you can use `apartment.address`, (which also includes apartment stair)

Reply:
Yes, I think we can filter out just the apartments we need from the database, I'll refactor.

One note on the current query, it only gets the "street_address", "apartment_number", "share_number_start", and "share_number_end" as a QuerySet containing a dict with `values()`, not the whole object, so the memory required is not as bad.

Reply:
This change is in the latest commit